wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsnext/WorkspaceNextObjectsRetriever.java

```diff
@@ -0,0 +1,403 @@
+      throws InfrastructureException {
+    Map<String, CheService> services = new HashMap<>();
+    for (CheFeature feature : features) {
+      for (CheServiceReference serviceReference : feature.getSpec().getServices()) {
+        String serviceName = serviceReference.getName();
+        String serviceVersion = serviceReference.getVersion();
+        String key = serviceName + '/' + serviceVersion;
+        if (!services.containsKey(key)) {
+          CheService service = getService(serviceName, serviceVersion);
+          services.put(key, service);
+        }
+      }
+    }
+    return services.values();
+  }
+
+  private boolean isParameter(EnvVar envVar) {
+    return PARAMETER_ENV_VAR_VALUE.matcher(envVar.getValue()).matches();
+  }
+
+  /**
+   * Returns map where key is concatenation of service name, version and parameter name separated by
+   * slash symbol. < serviceName/serviceVersion/${parameter} > to < parameterValue >
+   */
+  private Map<String, List<String>> getServicesParameters(Collection<CheFeature> features) {
+    Map<String, List<String>> parameters = new HashMap<>();
+    for (CheFeature feature : features) {
+      for (CheServiceReference serviceReference : feature.getSpec().getServices()) {
+        if (serviceReference.getParameters().isEmpty()) {
+          continue;
+        }
+
+        String serviceName = serviceReference.getName();
+        String serviceVersion = serviceReference.getVersion();
+
+        for (CheServiceParameter cheServiceParameter : serviceReference.getParameters()) {
+          // add dollar sign and curly brackets because parameter is easier to find with these signs
+          // in the map keys
+          String parameterKey =
+              serviceName + "/" + serviceVersion + "/${" + cheServiceParameter.getName() + "}";
+          List<String> cheServiceParameters =
+              parameters.computeIfAbsent(parameterKey, key -> new ArrayList<>());
+          cheServiceParameters.add(cheServiceParameter.getValue());
+        }
+      }
+    }
+    return parameters;
+  }
+
+  private CheService getService(String serviceName, String serviceVersion)
+      throws InfrastructureException {
+    try {
+      URI getServiceURI =
+          featureApi.clone().path("service").path(serviceName).path(serviceVersion).build();
+
+      CheService service = getBody(getServiceURI, CheService.class);
+      validateService(service, serviceName, serviceVersion);
+      return service;
+    } catch (IllegalArgumentException | UriBuilderException e) {
+      throw new InternalInfrastructureException(
+          format("Service %s/%s retrieval failed", serviceName, serviceVersion));
+    } catch (IOException e) {
+      throw new InfrastructureException(
+          format(
+              "Error occurred on retrieval of service %s. Error: %s",
+              serviceName + '/' + serviceVersion, e.getMessage()));
+    }
+  }
+
+  private void validateFeature(CheFeature feature, String name, String version)
```

Comment:
Just proposal. Maybe it would be better to move validation private methods to new `Validator` class to make this class a bit smaller.

Reply:
I would not want to create a separate class for that. At least for now.